### ISSUE TITLE
Client relationships enhancement

### ISSUE THREAD

**Issue by ghstahl** [IdentityServer/IdentityServer4#847]
### Initial Problem
In some case the downstream usage of the refresh_tokens should not be require to provide a client_secret.

The following is my test client that I use to create the refresh_token.
```
var resourceOwnerClient = new Client
{
      ClientId = "resource-owner-client",
      AllowedGrantTypes = GrantTypes.ResourceOwnerPassword,
      AllowOfflineAccess = true,
      RefreshTokenUsage = TokenUsage.OneTimeOnly,
      ClientSecrets =
      {
           new Secret("secret".Sha256())
      },
      AllowedScopes = { "arbitrary"}
};
```
I successfully can get a refresh_token, and successfully get another downstream, but only when I produce the same clientId/clientSecret.  _All as expected and documented_.

### What IdentityServer4 code says
```
class ClientSecretValidator 
{
    public async Task<ClientSecretValidationResult> ValidateAsync(HttpContext context)
    {
            .....
            if (!client.RequireClientSecret || client.IsImplicitOnly())
            {
                _logger.LogDebug("Public Client - skipping secret validation success");
            }
           else
            {
                 .....
            }
            .....
    }
}
```
Neither of which fits here as I want the client that created this to have a client_secret.

### Before I go forking

I think in all cases a relationship has to be established between 2 clients;
```
var resourceOwnerClient = new Client
{
      ClientId = "public-resource-owner-client",
      InheritedClient= "resource-owner-client",
      RequireClientSecret = false
};
```
#### Client Morphing during a grant_type=refresh_token call
```
class ClientSecretValidator 
{
    public async Task<ClientSecretValidationResult> ValidateAsync(HttpContext context)
    {
            .....
            client.MorphToParentIfInherited();
            if (!client.RequireClientSecret 
                || client.IsImplicitOnly())
            {
                _logger.LogDebug("Public Client - skipping secret validation success");
            }
           else
            {
                 .....
            }
            .....
    }
}
```

1. I create a refresh_token using resource-owner-client
2. I give out refresh_token and what client to use (public-resource-owner-client) to the public
3. public apps will be coming in with 
```
grant_type=refresh_token&refresh_token={refresh_token}&client_id=public-resource-owner-client
```
4. IdentityServer4 will for this request morph the public-resource-owner-client into resource-owner-client taking all the settings that make sense with it.  _**In this case changing the RequireClientSecret from a true to a false.**_  

#### Using the ClientSecret of the parent at the time or creation
In this possible solution, I would use the **public-resource-owner-client** to create the offline_access token using the ClientSecret of **resource-owner-client**.  **_I prefer this approach because it would result in the produced access_token to contain "client_id": "public-resource-owner-client"._**  

**Question**: I haven't played with anything where RequireClientSecret=false.  If the access_token, and by association the refresh_token contain the "client_id" which created them, why is it required to be passed as what appears to be a redundant argument?  I would say that at a minimum I should need nothing more than the following;
```
grant_type=refresh_token&refresh_token={refresh_token}
vs
grant_type=refresh_token&refresh_token={refresh_token}&client_id=public-resource-owner-client
```

### Finally
In your opinion is this a reasonable feature that IdentityServer4 should take up, provided it doesn't break the specification?

If it does break the spec, is it still reasonable to allow it to occur through abstraction thus pushing the responsibility of breaking the spec to me?

**Comment by leastprivilege** [comment]
Yea - all this morphing and inheriting is nothing I can see in our core.

Why don't you use an extension grant to implement your scenario?

**Comment by ghstahl** [comment]
OK, given that would you consider exposing a higher level way of dealing with refresh_tokens.  
https://identityserver4.readthedocs.io/en/release/topics/extension_grants.html
In your extensions grant example of DelegationGrantValidator you inject ITokenValidator, which only deals with access_tokens and identity_tokens.  From the research I have done, it would nice to have what exists in ITokenRequestValidator.  I tried injecting ITokenRequestValidator and got DI exceptions, but I would guess you might not want me to use that one as at that point I am already being called from your instance of that object. 

I am a good deal down the path of getting something working, but I had to lift a lot of your code out of TokenRequestValidator.ValidateRefreshTokenRequestAsync(NameValueCollection parameters) and injecting ITokenResponseGenerator.  

I am basically reproducing a refresh_token version of TokenRequestValidator, which is starting to feel wrong.

Since your stuff created the offline_access to begin with, I was hoping for you to expose a way to refresh it.
Thanks

**Comment by leastprivilege** [comment]
Maybe it would help if you explain what you really want to achieve and why. I still don't get it.

**Comment by ghstahl** [comment]
1. I have a client that is configured as follows;
```
var resourceOwnerClient = new Client
{
      ClientId = "resource-owner-client",
      AllowedGrantTypes = GrantTypes.ResourceOwnerPassword,
      AllowOfflineAccess = true,
      RefreshTokenUsage = TokenUsage.OneTimeOnly,
      ClientSecrets =
      {
           new Secret("secret".Sha256())
      },
      AllowedScopes = { "arbitrary"}
};
```
2. I am able to use this client to give me an access_token, refresh_token, etc.  using the offline_access scope.
3. I am able to use this refresh_token to refresh, provided I pass all the necessary arguement.  clientId, clientSecret, refresh_token, etc.   
4. Everything up to this point is as expected, documented, all good.
5. I am giving this refresh_token, to an untrusted endpoint.   This is very similar to what is described here.  https://identityserver4.readthedocs.io/en/release/topics/extension_grants.html
6. The untrusted endpoint is a mobile app.
7. I would like the Mobil App to have the ability to use the refresh_token, as I did in step 3 to get new tokens, however I will not be sharing my client_secret with them.

Going down the extension_grants path, I theorized that I should be able to do the following;
```
http://localhost:7791/connect/token POST
grant_type=public_refresh_token&refresh_token={token created by resource-owner-client}&client_id=public-resource-owner-client
```
8. I created another client, that by convention was named based upon another client.  Please note the AllowedGrantTypes and RequireClientSecret  in "public-resource-owner-client"
```
new Client
            {
                ClientId = "public-resource-owner-client",
                AllowedGrantTypes = GrantTypes.List("public_refresh_token"),
                RequireClientSecret = false,
                AllowOfflineAccess = true
            }
```
9. So work to date.  I am able to get the public_refresh_token request and successfully validate the refresh_token against the client that created it, but I am stopping short of finishing this work until I get some sort of validation from you.   I am finding that I am copying most your code from TokenRequestValidator to get as far as I did. 
```
My PublicRefreshTokenExtensionGrantValidator  
is staring to look like a mini version of your TokenRequestValidator.

 public class PublicRefreshTokenExtensionGrantValidator : IExtensionGrantValidator
    {
        private readonly ILogger<PublicRefreshTokenExtensionGrantValidator> _logger;
        private readonly IEventService _events;
        private readonly IdentityServerOptions _options;
        private readonly IRefreshTokenStore _refreshTokenStore;
        private readonly IProfileService _profile;
        private readonly IClientStore _clientStore;
        private readonly ITokenResponseGenerator _tokenResponseGenerator;

        private ValidatedTokenRequest _validatedRequest;
        private const string PrependPublic = "public-";
        private const int PrependPublicIndex = 7;
        public PublicRefreshTokenExtensionGrantValidator(
            IdentityServerOptions options,
            IRefreshTokenStore refreshTokenStore,
            ITokenResponseGenerator tokenResponseGenerator,
            IProfileService profile,
            IClientStore clientStore,
            IEventService events,
            ILogger<PublicRefreshTokenExtensionGrantValidator> logger)
        {
            _logger = logger;
            _events = events;
            _options = options;
            _refreshTokenStore = refreshTokenStore;
            _profile = profile;
            _clientStore = clientStore;
            _tokenResponseGenerator = tokenResponseGenerator;
        }
}
```

**Comment by leastprivilege** [comment]
And WHY are you doing 3 (and expect 7 to work?)

**Comment by ghstahl** [comment]
3 is works, but it is not my final use case.  However, 3 is a critical piece for me and has to have occurred as is.  Only later, with proper configuration, 7 can occur as described. 

7 is my ask.

7 currently does not work, BUT given your approval I expect it to work.  

The reason I expect 7 to work is that I have used other OAuth 2.0 systems where this was allowed.  I had to configure it, but nevertheless allowed.

The benefit for me, if you agree that this is valid, is that IdentityServer4 would take up the work to allow it.  I am more than willing to help with this.

**So I got it working using extension grants, but I feel a bit dirty about it ;)**
https://gist.github.com/ghstahl/6c22ff4d82e686e52b5334c9d2424b3e

My extension does the following;
1. Validate a refresh_token that was created by another client
2. refresh it.  I don't like this step, but had to do it under the current IdentityServer4 constraints.
3. Turn that refresh result into a custom response, and call it inner_response.

if you Jwt.io this you can see the clients in play.
````
public-resource-owner-client response, which wraps the resource-owner-client response
{
    "access_token":  "eyJhbGciOiJSUzI1NiIsImtpZCI6IjhlNDJiNTI0OTZmNTM5Nzk2MjA5YzNjZDAyOWJmM2FhIiwidHlwIjoiSldUIn0.eyJuYmYiOjE0ODgzODQ0MjgsImV4cCI6MTQ4ODM4ODAyOCwiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo3NzkxIiwiYXVkIjpbImh0dHA6Ly9sb2NhbGhvc3Q6Nzc5MS9yZXNvdXJjZXMiLCJhcmJpdHJhcnkiXSwiY2xpZW50X2lkIjoicHVibGljLXJlc291cmNlLW93bmVyLWNsaWVudCIsInNjb3BlIjpbImFyYml0cmFyeSJdfQ.Qe9yjSJN6r1oHRPecBpxA2VdyhJMvZX5gfH6495lQb8-m6kkuE_0AMmFsIpexsvaHLtHTYztZvRgiDXVzb49eqwBUPrmsOrXidzQTiEKfHZXfnPBAxOaWyE2EXq418_Uv1cU3z6OlmQjGg7CEsrcyWx68xlwZ3wDxnwLCri6eo6pF28ClD2vr_IEPyeym5ej8ybgIApcR6uzzRdHAiKXiSzMq2WnSpQED6eJ_Mj0Leg-QJd3_3eo42GexeFgu109aiAB697iCVnAA5jpXo5NsSp1wXKgHQVdNGIfhi39LuAryl1vn5D6zKWGiNmb-u-X9YQlYsR8miCewCNJRtZmYg",
    "expires_in": 3600,
    "token_type": "Bearer",
    "inner_response": {
        "access_token": "eyJhbGciOiJSUzI1NiIsImtpZCI6IjhlNDJiNTI0OTZmNTM5Nzk2MjA5YzNjZDAyOWJmM2FhIiwidHlwIjoiSldUIn0.eyJuYmYiOjE0ODgzODQ0MjMsImV4cCI6MTQ4ODM4ODAyMywiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo3NzkxIiwiYXVkIjpbImh0dHA6Ly9sb2NhbGhvc3Q6Nzc5MS9yZXNvdXJjZXMiLCJhcmJpdHJhcnkiXSwiY2xpZW50X2lkIjoicmVzb3VyY2Utb3duZXItY2xpZW50Iiwic3ViIjoicmF0IiwiYXV0aF90aW1lIjoxNDg4Mzg0NDA0LCJpZHAiOiJsb2NhbCIsIm5hZ3VpZCI6IjEyMzRhYmNkIiwiSW4iOiJGbGFtZXMiLCJzY29wZSI6WyJhcmJpdHJhcnkiLCJvZmZsaW5lX2FjY2VzcyIsIkEiLCJxdWljayIsImJyb3duIiwiZm94Il0sImFtciI6WyJwd2QiXX0.ot1nMhbyTwmOpG6G6ghFkG9PrZCtXO_4nonHb0NEaDLXmI7aBcXEx9WiYcugz_J0-AHkPjCP8n67VI1Z4oIp7iyD7o_37ydFbGGnc2V4NtD_GzwV_rqtk6rtfNKKQpkOU9MTI_KJzoVGjgrYDjyoK8iFm828jzuRBr2xmLdvakL1-fwitGeWxU-NJPrzJA8AQothhLvOjTqQem6-O8LBM7wVAOXQGirfSghfvz61D2hv8qp0k1VmxL1XUz_ESy5Bp0Wg8QxkupS1lsg_6RdAqNzOEKyZpdZT_j3Wy2rg3-c8HLokK_IOm9Ek6xr9plf60FfqjiYmLDXGJIXSxnYkXg",
        "expires_in": 3600,
        "token_type": "Bearer",
        "refresh_token": "b1cca6c78cefcef17259b9c9636c9beae05238f96d1a36f07dacdd983beffe7f"
    }
}
````

ideally I would have liked the response to be this;
```
resource-owner-client response
{
    "access_token": "eyJhbGciOiJSUzI1NiIsImtpZCI6IjhlNDJiNTI0OTZmNTM5Nzk2MjA5YzNjZDAyOWJmM2FhIiwidHlwIjoiSldUIn0.eyJuYmYiOjE0ODgzODQ0MjMsImV4cCI6MTQ4ODM4ODAyMywiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo3NzkxIiwiYXVkIjpbImh0dHA6Ly9sb2NhbGhvc3Q6Nzc5MS9yZXNvdXJjZXMiLCJhcmJpdHJhcnkiXSwiY2xpZW50X2lkIjoicmVzb3VyY2Utb3duZXItY2xpZW50Iiwic3ViIjoicmF0IiwiYXV0aF90aW1lIjoxNDg4Mzg0NDA0LCJpZHAiOiJsb2NhbCIsIm5hZ3VpZCI6IjEyMzRhYmNkIiwiSW4iOiJGbGFtZXMiLCJzY29wZSI6WyJhcmJpdHJhcnkiLCJvZmZsaW5lX2FjY2VzcyIsIkEiLCJxdWljayIsImJyb3duIiwiZm94Il0sImFtciI6WyJwd2QiXX0.ot1nMhbyTwmOpG6G6ghFkG9PrZCtXO_4nonHb0NEaDLXmI7aBcXEx9WiYcugz_J0-AHkPjCP8n67VI1Z4oIp7iyD7o_37ydFbGGnc2V4NtD_GzwV_rqtk6rtfNKKQpkOU9MTI_KJzoVGjgrYDjyoK8iFm828jzuRBr2xmLdvakL1-fwitGeWxU-NJPrzJA8AQothhLvOjTqQem6-O8LBM7wVAOXQGirfSghfvz61D2hv8qp0k1VmxL1XUz_ESy5Bp0Wg8QxkupS1lsg_6RdAqNzOEKyZpdZT_j3Wy2rg3-c8HLokK_IOm9Ek6xr9plf60FfqjiYmLDXGJIXSxnYkXg",
    "expires_in": 3600,
    "token_type": "Bearer",
    "refresh_token": "b1cca6c78cefcef17259b9c9636c9beae05238f96d1a36f07dacdd983beffe7f"
}
```

I get that one might be concerned that I am passing back a response from another client, but I think it is just delegation and the security issues are accounted for.

Also, I used all your code to achieve this, so I was hoping you would expose the parts I lifted without change be exposed by IdentityServer4.  Mainly to validate a refresh_token.

-H

**Comment by leastprivilege** [comment]
I am curious - what’s the business use case around this technical
requirement?


-------
Dominick Baier

On 1 March 2017 at 17:32:25, Herb Stahl (notifications@github.com) wrote:

3 is works, but it is not my final use case. However, 3 is a critical piece
for me and has to have occurred as is. Only later, with proper
configuration, 7 can occur as described.

7 is my ask.

7 currently does not work, BUT given your approval I expect it to work.

The reason I expect 7 to work is that I have used other OAuth 2.0 systems
where this was allowed. I had to configure it, but nevertheless allowed.

The benefit for me, if you agree that this is valid, is that
IdentityServer4 would take up the work to allow it. I am more than willing
to help with this.

*So I got it working using extension grants, but I feel a bit dirty about
it ;)*
https://gist.github.com/ghstahl/6c22ff4d82e686e52b5334c9d2424b3e

My extension does the following;

   1. Validate a refresh_token that was created by another client
   2. refresh it. I don't like this step, but had to do it under the
   current IdentityServer4 constraints.
   3. Turn that refresh result into a custom response, and call it
   inner_response.

if you Jwt.io this you can see the clients in play.

public-resource-owner-client response, which wraps the
resource-owner-client response
{
    "access_token":
"eyJhbGciOiJSUzI1NiIsImtpZCI6IjhlNDJiNTI0OTZmNTM5Nzk2MjA5YzNjZDAyOWJmM2FhIiwidHlwIjoiSldUIn0.eyJuYmYiOjE0ODgzODQ0MjgsImV4cCI6MTQ4ODM4ODAyOCwiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo3NzkxIiwiYXVkIjpbImh0dHA6Ly9sb2NhbGhvc3Q6Nzc5MS9yZXNvdXJjZXMiLCJhcmJpdHJhcnkiXSwiY2xpZW50X2lkIjoicHVibGljLXJlc291cmNlLW93bmVyLWNsaWVudCIsInNjb3BlIjpbImFyYml0cmFyeSJdfQ.Qe9yjSJN6r1oHRPecBpxA2VdyhJMvZX5gfH6495lQb8-m6kkuE_0AMmFsIpexsvaHLtHTYztZvRgiDXVzb49eqwBUPrmsOrXidzQTiEKfHZXfnPBAxOaWyE2EXq418_Uv1cU3z6OlmQjGg7CEsrcyWx68xlwZ3wDxnwLCri6eo6pF28ClD2vr_IEPyeym5ej8ybgIApcR6uzzRdHAiKXiSzMq2WnSpQED6eJ_Mj0Leg-QJd3_3eo42GexeFgu109aiAB697iCVnAA5jpXo5NsSp1wXKgHQVdNGIfhi39LuAryl1vn5D6zKWGiNmb-u-X9YQlYsR8miCewCNJRtZmYg",
    "expires_in": 3600,
    "token_type": "Bearer",
    "inner_response": {
        "access_token":
"eyJhbGciOiJSUzI1NiIsImtpZCI6IjhlNDJiNTI0OTZmNTM5Nzk2MjA5YzNjZDAyOWJmM2FhIiwidHlwIjoiSldUIn0.eyJuYmYiOjE0ODgzODQ0MjMsImV4cCI6MTQ4ODM4ODAyMywiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo3NzkxIiwiYXVkIjpbImh0dHA6Ly9sb2NhbGhvc3Q6Nzc5MS9yZXNvdXJjZXMiLCJhcmJpdHJhcnkiXSwiY2xpZW50X2lkIjoicmVzb3VyY2Utb3duZXItY2xpZW50Iiwic3ViIjoicmF0IiwiYXV0aF90aW1lIjoxNDg4Mzg0NDA0LCJpZHAiOiJsb2NhbCIsIm5hZ3VpZCI6IjEyMzRhYmNkIiwiSW4iOiJGbGFtZXMiLCJzY29wZSI6WyJhcmJpdHJhcnkiLCJvZmZsaW5lX2FjY2VzcyIsIkEiLCJxdWljayIsImJyb3duIiwiZm94Il0sImFtciI6WyJwd2QiXX0.ot1nMhbyTwmOpG6G6ghFkG9PrZCtXO_4nonHb0NEaDLXmI7aBcXEx9WiYcugz_J0-AHkPjCP8n67VI1Z4oIp7iyD7o_37ydFbGGnc2V4NtD_GzwV_rqtk6rtfNKKQpkOU9MTI_KJzoVGjgrYDjyoK8iFm828jzuRBr2xmLdvakL1-fwitGeWxU-NJPrzJA8AQothhLvOjTqQem6-O8LBM7wVAOXQGirfSghfvz61D2hv8qp0k1VmxL1XUz_ESy5Bp0Wg8QxkupS1lsg_6RdAqNzOEKyZpdZT_j3Wy2rg3-c8HLokK_IOm9Ek6xr9plf60FfqjiYmLDXGJIXSxnYkXg",
        "expires_in": 3600,
        "token_type": "Bearer",
        "refresh_token":
"b1cca6c78cefcef17259b9c9636c9beae05238f96d1a36f07dacdd983beffe7f"
    }
}

ideally I would have liked the response to be this;

resource-owner-client response
{
    "access_token":
"eyJhbGciOiJSUzI1NiIsImtpZCI6IjhlNDJiNTI0OTZmNTM5Nzk2MjA5YzNjZDAyOWJmM2FhIiwidHlwIjoiSldUIn0.eyJuYmYiOjE0ODgzODQ0MjMsImV4cCI6MTQ4ODM4ODAyMywiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo3NzkxIiwiYXVkIjpbImh0dHA6Ly9sb2NhbGhvc3Q6Nzc5MS9yZXNvdXJjZXMiLCJhcmJpdHJhcnkiXSwiY2xpZW50X2lkIjoicmVzb3VyY2Utb3duZXItY2xpZW50Iiwic3ViIjoicmF0IiwiYXV0aF90aW1lIjoxNDg4Mzg0NDA0LCJpZHAiOiJsb2NhbCIsIm5hZ3VpZCI6IjEyMzRhYmNkIiwiSW4iOiJGbGFtZXMiLCJzY29wZSI6WyJhcmJpdHJhcnkiLCJvZmZsaW5lX2FjY2VzcyIsIkEiLCJxdWljayIsImJyb3duIiwiZm94Il0sImFtciI6WyJwd2QiXX0.ot1nMhbyTwmOpG6G6ghFkG9PrZCtXO_4nonHb0NEaDLXmI7aBcXEx9WiYcugz_J0-AHkPjCP8n67VI1Z4oIp7iyD7o_37ydFbGGnc2V4NtD_GzwV_rqtk6rtfNKKQpkOU9MTI_KJzoVGjgrYDjyoK8iFm828jzuRBr2xmLdvakL1-fwitGeWxU-NJPrzJA8AQothhLvOjTqQem6-O8LBM7wVAOXQGirfSghfvz61D2hv8qp0k1VmxL1XUz_ESy5Bp0Wg8QxkupS1lsg_6RdAqNzOEKyZpdZT_j3Wy2rg3-c8HLokK_IOm9Ek6xr9plf60FfqjiYmLDXGJIXSxnYkXg",
    "expires_in": 3600,
    "token_type": "Bearer",
    "refresh_token":
"b1cca6c78cefcef17259b9c9636c9beae05238f96d1a36f07dacdd983beffe7f"
}

I get that one might be concerned that I am passing back a response from
another client, but I think it is just delegation and the security issues
are accounted for.

Also, I used all your code to achieve this, so I was hoping you would
expose the parts I lifted without change be exposed by IdentityServer4.
Mainly to validate a refresh_token.

-H

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/IdentityServer/IdentityServer4/issues/847#issuecomment-283391931>,
or mute the thread

**Comment by ghstahl** [comment]
Imagine a mobile app, that by a lottery system gives out access to resources.
No user needed.
First 100 in get access, and access is for 1 day.
Following good practices, I wanted my access_token to be short lived (10 minutes), and refresh_token to be long lived (1 day).
All this stuff is configurable.

I wanted my MobileApps to use a standard way to get tokens, a la IdentityServer4.  I didn't want to write a poor mans delegation endpoint that now is yet another standard.

We really wanted a client_credential flow with offline_access, where our backed using its clientId/clientSecret would create a token.  This token was given to our MobilApp.  We also wanted the client to be given a refresh_token, so they could refresh this particular access_token.  client_credential flow doesn't allow offline_access, and in this case there is no user in the picture.

So I went down the path of Resource_Owner, which does allow offline_access.
This lead me to building out something where I could create a Resource_Owner offline_access token that simply accepted any user without question.  Its basically a glorified client_credentials flow, which now allows offline_access.

Then I needed to let these unsecured clients refresh the tokens, without a client_secret.
That brought in an enhanced_grant flow.

**Comment by leastprivilege** [comment]
This seems all way too complicated for me.

Have you considered giving out reference tokens? This way you could make them longer lived - but still are able to revoke them at any point in time.

**Comment by ghstahl** [comment]
Good point, and I have that at my disposal.  The design pattern that has been most reliable for me has been to put in rules on the edge, as opposed to writing code that say sweeps a backend database going on a revocation spree.

My choices are based upon what constraints any given organization imposes on me, and admin access to databases to revoke references in a sweep may actually be banned.  So I find other ways more tolerably accepted, and have turned out to be more reliable.

Lets say I have 50Million+ endpoints that I handed out refresh_tokens.  In a lot of our cases an access_token is for 1 day, with a refresh_token for 30 days.  As these endpoints come trickling in trying to refresh, I can apply my rules right there and on-demand, I can also change my mind and only a small portion of the population is affected.

To your(IdentityServer4) credit,  it resulted in very little code of which most of it I copied from you :).

Now that it is in place, its usage is quite simple, and really flexible.

Did you give any thought on exposing the ability to validate a refresh_token?  Perhaps a tiny little new method in ITokenValidator?
